R/formula.R

```diff
@@ -0,0 +1,407 @@
+#' @return a list including the updated formula and data
+subset_formula_and_data <- function(formula, terms, data, y=NULL, split_formula=FALSE) {
+  formula <- subset_formula(formula, terms)
+  tt <- extract_terms_response(formula)
+  response_name <- tt$response
+
+  if (!is.null(ncol(y)) && ncol(y) > 1) {
+    response_cols <- unlist(lapply(1:ncol(y), function(n)
+      paste0(".", response_name, ".", n)))
+    if (!split_formula) {
+      response_vector <- paste0("cbind(", paste(response_cols, collapse=", "), ")")
+      formula <- update(formula, paste0(response_vector, " ~ ."))
+    } else {
+      formula <- lapply(response_cols, function(response)
+        update(formula, paste0(response, " ~ .")))
+    }
+    data <- data.frame(y, data[, colnames(data) != response_name])
+    colnames(data)[1:ncol(y)] <- response_cols
+  }
+
+  return(nlist(formula, data))
+}
+
+#' Subsets a formula by the given terms.
+#' @param formula A formula for a valid model.
+#' @param terms A vector of terms to subset from the right hand side.
+#' @return the updated formula
+subset_formula <- function(formula, terms) {
+  update(formula, paste0(". ~ ", paste(terms, collapse=" + ")))
+}
+
+#' Utility to count the number of terms in a given formula.
+#' @param subformula A formula for a valid model.
+#' @return the number of terms in the subformula.
+count_terms_in_subformula <- function(subformula) {
+  if (!inherits(subformula, "formula"))
+    subformula <- as.formula(paste0("~ ", subformula))
+  tt <- extract_terms_response(subformula)
+  ind_interaction_terms <- length(tt$individual_terms) + length(tt$interaction_terms)
+  group_terms <- sum(unlist(lapply(tt$group_terms, count_terms_in_group_term)))
+  ind_interaction_terms + group_terms
+}
+
+#' Utility to count the number of terms in a given group term.
+#' @param term A group term as a string.
+#' @return the number of terms in the group term.
+count_terms_in_group_term <- function(term) {
+  term <- gsub("[\\(\\)]", "", flatten_group_terms(term))
+
+  chunks <- strsplit(term,  "[ ]*\\|([^\\|]*\\||)[ ]*")[[1]]
+  lhs <- as.formula(paste0("~", chunks[1]))
+  tt <- extact_terms_response(lhs)
+  variables <- c(tt$individual_terms, tt$interaction_terms)
+
+  ## split possible interaction terms
+  int_t <- grepl(":", variables)
+  int_v <- variables[int_t]
+  lin_v <- variables[!int_t]
+
+  if ("0" %in% variables) {
+    group_intercept <- FALSE
+    variables <- setdiff(variables, "0")
+  } else {
+    group_intercept <- TRUE
+    if (!("1" %in% variables))
+      variables <- c(variables, "1")
+  }
+
+  length(variables)
+}
+
+#' Given a list of formulas, sort them by number of terms in them.
+#' @param submodels A list of models' formulas.
+#' @return a sorted list of submodels by included terms.
+sort_submodels_by_size <- function(submodels) {
+  size <- lapply(submodels, count_terms_in_subformula)
+  df <- as.data.frame(cbind(submodels=submodels, size=unlist(size)))
+  df$size <- as.numeric(df$size)
```

Comment:
Can we avoid creating a matrix and then a data.frame and create a data.frame directly? Then we can also get rid of the `as.numeric`.

Reply:
fixed